File: 2015/python/12.py
# Advent of Code 2015, day 12
# (c) aichingert

import json

def part1(data, valid):
    ans = ""
    for ch in data:
        if ch in valid:
            ans += ch
        else:
            ans += ' '
    return sum([int(i) for i in ans.split()])

def rem_red(js):
    if isinstance(js,dict):
        for k,v in js.items():
            if k == "red" or v == "red":
                return None
        ans = dict()
        for k, v in js.items():
            ans[k] = rem_red(v)
        return ans
    elif isinstance(js,list):
        ans = []
        for v in js:
            ans.append(rem_red(v))
        return ans
    else:
        return js

def part2(data, valid):
    return part1(str(rem_red(json.loads(data.strip()))), valid)


inp = open("../input/12").read()
valid = "-0123456789"

print("Part 1:",part1(inp, valid))
print("Part 2:",part2(inp, valid))

File: 2015/rust/12.rs
// Advent of Code 2015, day 12
// (c) aichingert

const V: [char;11] = ['-','0','1','2','3','4','5','6','7','8','9'];

fn part1(ch: &[char]) -> i32 {
    let mut ans = 0i32;
    let mut cur = String::new();
    
    for i in 0..ch.len() {
        if V.contains(&ch[i]) {
            cur.push(ch[i]);
        } else {
            cur.push(' ');
        }
    }

    cur.split(' ').filter(|s| !s.is_empty()).map(|s| s.parse::<i32>().unwrap()).sum::<i32>()
}

fn part2() {}

fn main() {
    let inp = std::fs::read_to_string("../input/12").unwrap();
    let inp = inp.trim().chars().collect::<Vec<char>>();

    println!("Part 1: {}", part1(&inp));
}
